File: src/main/java/io/github/takenoha/towerdefense/tactical/TacticalEffectSnapshotImpl.java
package io.github.takenoha.towerdefense.tactical;

import io.github.takenoha.towerdefense.domain.TowerType;
import java.util.EnumMap;
import java.util.Map;
import java.util.Objects;

/** Immutable, pre-aggregated tactical effects for one active defense. */
public final class TacticalEffectSnapshotImpl implements TacticalEffectSnapshot {
    private final EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> damage;
    private final EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> attackInterval;
    private final EnumMap<TowerType, Double> rangeAdd;
    private final EnumMap<TowerType, Double> areaRadius;
    private final EnumMap<TowerType, Integer> chainCount;
    private final EnumMap<TowerType, Double> slowStrength;
    private final EnumMap<TowerType, Double> burnDuration;
    private final double supportBuff;
    private final double repairCost;
    private final double towerDamageTaken;

    TacticalEffectSnapshotImpl(
            EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> damage,
            EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> attackInterval,
            EnumMap<TowerType, Double> rangeAdd,
            EnumMap<TowerType, Double> areaRadius,
            EnumMap<TowerType, Integer> chainCount,
            EnumMap<TowerType, Double> slowStrength,
            EnumMap<TowerType, Double> burnDuration,
            double supportBuff,
            double repairCost,
            double towerDamageTaken) {
        this.damage = copyConditional(damage);
        this.attackInterval = copyConditional(attackInterval);
        this.rangeAdd = new EnumMap<>(Objects.requireNonNull(rangeAdd, "rangeAdd"));
        this.areaRadius = new EnumMap<>(Objects.requireNonNull(areaRadius, "areaRadius"));
        this.chainCount = new EnumMap<>(Objects.requireNonNull(chainCount, "chainCount"));
        this.slowStrength = new EnumMap<>(Objects.requireNonNull(slowStrength, "slowStrength"));
        this.burnDuration = new EnumMap<>(Objects.requireNonNull(burnDuration, "burnDuration"));
        this.supportBuff = positiveFiniteOrNeutral(supportBuff);
        this.repairCost = positiveFiniteOrNeutral(repairCost);
        this.towerDamageTaken = positiveFiniteOrNeutral(towerDamageTaken);
    }

    @Override
    public double damageMultiplier(TowerType type, TacticalTargetContext target) {
        return conditionalValue(damage, type, target);
    }

    @Override
    public double attackIntervalMultiplier(TowerType type, TacticalTargetContext target) {
        return conditionalValue(attackInterval, type, target);
    }

    @Override
    public double rangeAdd(TowerType type) {
        return rangeAdd.getOrDefault(Objects.requireNonNull(type, "type"), 0.0d);
    }

    @Override
    public double areaRadiusMultiplier(TowerType type) {
        return areaRadius.getOrDefault(
                Objects.requireNonNull(type, "type"), 1.0d);
    }

    @Override
    public int chainCountAdd(TowerType type) {
        return chainCount.getOrDefault(Objects.requireNonNull(type, "type"), 0);
    }

    @Override
    public double slowStrengthMultiplier(TowerType type) {
        return slowStrength.getOrDefault(
                Objects.requireNonNull(type, "type"), 1.0d);
    }

    @Override
    public double burnDurationMultiplier(TowerType type) {
        return burnDuration.getOrDefault(
                Objects.requireNonNull(type, "type"), 1.0d);
    }

    @Override
    public double supportBuffMultiplier() {
        return supportBuff;
    }

    @Override
    public double repairCostMultiplier() {
        return repairCost;
    }

    @Override
    public double towerDamageTakenMultiplier() {
        return towerDamageTaken;
    }

    private static EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> copyConditional(
            EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> source) {
        Objects.requireNonNull(source, "source");
        EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> copy =
                new EnumMap<>(TowerType.class);
        for (TowerType type : TowerType.values()) {
            EnumMap<TacticalTargetCondition, Double> row = source.get(type);
            copy.put(
                    type,
                    row == null
                            ? new EnumMap<>(TacticalTargetCondition.class)
                            : new EnumMap<>(row));
        }
        return copy;
    }

    private static double conditionalValue(
            Map<TowerType, ? extends Map<TacticalTargetCondition, Double>> values,
            TowerType type,
            TacticalTargetContext target) {
        Objects.requireNonNull(type, "type");
        Objects.requireNonNull(target, "target");
        Map<TacticalTargetCondition, Double> row = values.get(type);
        if (row == null || row.isEmpty()) {
            return 1.0d;
        }
        double result = 1.0d;
        for (Map.Entry<TacticalTargetCondition, Double> entry : row.entrySet()) {
            if (matches(entry.getKey(), target)) {
                result = safeProduct(result, entry.getValue());
            }
        }
        return positiveFiniteOrNeutral(result);
    }

    private static boolean matches(
            TacticalTargetCondition condition,
            TacticalTargetContext target) {
        return switch (condition) {
            case NONE -> true;
            case CORE_BELOW_50_PERCENT -> target.coreBelowHalf();
            case CORE_BELOW_30_PERCENT -> target.coreBelowThirtyPercent();
            case BOSS -> target.boss();
            case HIGH_HP -> target.targetHasHighHealth();
            case LOW_HP -> target.targetHasLowHealth();
            case SLOWED -> target.slowed();
            case BURNING -> target.burning();
        };
    }

    static double safeProduct(double left, double right) {
        if (!Double.isFinite(left) || !Double.isFinite(right)
                || left <= 0.0d || right <= 0.0d) {
            return 1.0d;
        }
        double result = left * right;
        return Double.isFinite(result) && result > 0.0d ? result : 1.0d;
    }

    static double positiveFiniteOrNeutral(double value) {
        return Double.isFinite(value) && value > 0.0d ? value : 1.0d;
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/tactical/TacticalEffectSnapshotImpl.kt
package io.github.takenoha.towerdefense.tactical

import io.github.takenoha.towerdefense.domain.TowerType
import java.util.EnumMap
import java.util.Objects

/** Immutable, pre-aggregated tactical effects for one active defense. */
class TacticalEffectSnapshotImpl internal constructor(
    damage: EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>>,
    attackInterval: EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>>,
    rangeAdd: EnumMap<TowerType, Double>,
    areaRadius: EnumMap<TowerType, Double>,
    chainCount: EnumMap<TowerType, Int>,
    slowStrength: EnumMap<TowerType, Double>,
    burnDuration: EnumMap<TowerType, Double>,
    supportBuff: Double,
    repairCost: Double,
    towerDamageTaken: Double,
) : TacticalEffectSnapshot {
    private val damage = copyConditional(damage)
    private val attackInterval = copyConditional(attackInterval)
    private val rangeAdd = EnumMap(
        Objects.requireNonNull(rangeAdd, "rangeAdd"),
    )
    private val areaRadius = EnumMap(
        Objects.requireNonNull(areaRadius, "areaRadius"),
    )
    private val chainCount = EnumMap(
        Objects.requireNonNull(chainCount, "chainCount"),
    )
    private val slowStrength = EnumMap(
        Objects.requireNonNull(slowStrength, "slowStrength"),
    )
    private val burnDuration = EnumMap(
        Objects.requireNonNull(burnDuration, "burnDuration"),
    )
    private val supportBuff = positiveFiniteOrNeutral(supportBuff)
    private val repairCost = positiveFiniteOrNeutral(repairCost)
    private val towerDamageTaken = positiveFiniteOrNeutral(towerDamageTaken)

    override fun damageMultiplier(
        type: TowerType,
        target: TacticalTargetContext,
    ): Double = conditionalValue(damage, type, target)

    override fun attackIntervalMultiplier(
        type: TowerType,
        target: TacticalTargetContext,
    ): Double = conditionalValue(attackInterval, type, target)

    override fun rangeAdd(type: TowerType): Double = rangeAdd.getOrDefault(
        Objects.requireNonNull(type, "type"),
        0.0,
    )

    override fun areaRadiusMultiplier(type: TowerType): Double = areaRadius.getOrDefault(
        Objects.requireNonNull(type, "type"),
        1.0,
    )

    override fun chainCountAdd(type: TowerType): Int = chainCount.getOrDefault(
        Objects.requireNonNull(type, "type"),
        0,
    )

    override fun slowStrengthMultiplier(type: TowerType): Double = slowStrength.getOrDefault(
        Objects.requireNonNull(type, "type"),
        1.0,
    )

    override fun burnDurationMultiplier(type: TowerType): Double = burnDuration.getOrDefault(
        Objects.requireNonNull(type, "type"),
        1.0,
    )

    override fun supportBuffMultiplier(): Double = supportBuff

    override fun repairCostMultiplier(): Double = repairCost

    override fun towerDamageTakenMultiplier(): Double = towerDamageTaken

    companion object {
        private fun copyConditional(
            source: EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>>,
        ): EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> {
            Objects.requireNonNull(source, "source")
            val copy = EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>>(
                TowerType::class.java,
            )
            for (type in TowerType.values()) {
                val row = source[type]
                copy[type] = row?.let {
                    EnumMap<TacticalTargetCondition, Double>(it)
                } ?: EnumMap(TacticalTargetCondition::class.java)
            }
            return copy
        }

        private fun conditionalValue(
            values: Map<TowerType, Map<TacticalTargetCondition, Double>>,
            type: TowerType,
            target: TacticalTargetContext,
        ): Double {
            Objects.requireNonNull(type, "type")
            Objects.requireNonNull(target, "target")
            val row = values[type]
            if (row == null || row.isEmpty()) {
                return 1.0
            }
            var result = 1.0
            for (entry in row.entries) {
                if (matches(entry.key, target)) {
                    result = safeProduct(result, entry.value)
                }
            }
            return positiveFiniteOrNeutral(result)
        }

        private fun matches(
            condition: TacticalTargetCondition,
            target: TacticalTargetContext,
        ): Boolean = when (condition) {
            TacticalTargetCondition.NONE -> true
            TacticalTargetCondition.CORE_BELOW_50_PERCENT -> target.coreBelowHalf()
            TacticalTargetCondition.CORE_BELOW_30_PERCENT -> target.coreBelowThirtyPercent()
            TacticalTargetCondition.BOSS -> target.boss()
            TacticalTargetCondition.HIGH_HP -> target.targetHasHighHealth()
            TacticalTargetCondition.LOW_HP -> target.targetHasLowHealth()
            TacticalTargetCondition.SLOWED -> target.slowed()
            TacticalTargetCondition.BURNING -> target.burning()
        }

        @JvmStatic
        fun safeProduct(left: Double, right: Double): Double {
            if (!left.isFinite() || !right.isFinite() || left <= 0.0 || right <= 0.0) {
                return 1.0
            }
            val result = left * right
            return if (result.isFinite() && result > 0.0) result else 1.0
        }

        private fun positiveFiniteOrNeutral(value: Double): Double =
            if (value.isFinite() && value > 0.0) value else 1.0
    }
}
